### PR TITLE
Remove PathNormalizationStrategy

### PR DESCRIPTION
**Pull request by wolfs**
We now have different fingerprinters - one for each path sensitivity.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/AbsolutePathInputNormalizer.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+package org.gradle.internal.fingerprint;
+
+import org.gradle.api.internal.changedetection.state.FileCollectionSnapshotter;
+import org.gradle.api.tasks.FileNormalizer;
+
+/**
+ * {@link FileNormalizer} that uses absolute paths for input files. The default.
+ */
+public interface AbsolutePathInputNormalizer extends FileCollectionSnapshotter, FileNormalizer {
+}
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/fingerprint/AbsolutePathInputNormalizer.java]: I don't think this makes sense to implement `FileCollectionSnapshotter`: we don't ever use it as such. Instead, we should make this public API as a `FileNormalizer`. Same for all the other new `FileNormalizer`s.

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultAbsolutePathInputNormalizer.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package org.gradle.internal.fingerprint.impl;
+
+import org.gradle.api.internal.cache.StringInterner;
+import org.gradle.api.internal.changedetection.state.FileSystemSnapshotter;
+import org.gradle.api.internal.changedetection.state.mirror.logical.AbsolutePathFingerprintingStrategy;
+import org.gradle.api.internal.file.collections.DirectoryFileTreeFactory;
+import org.gradle.api.tasks.FileNormalizer;
+import org.gradle.internal.fingerprint.AbsolutePathInputNormalizer;
+
+public class DefaultAbsolutePathInputNormalizer extends AbstractPathOnlyFileCollectionFingerprinter implements AbsolutePathInputNormalizer {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultAbsolutePathInputNormalizer.java]: Conversely, `FileCollectionFingerprinter` implementations shouldn't implement the corresponding normalizer. The relationship is instead defined via `getRegisteredType()`.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy [hunk 3]
```diff
@@ -18,12 +18,10 @@ package org.gradle.api.internal.changedetection.state
...
 import org.gradle.api.internal.changedetection.rules.AbstractTaskStateChangesTest
-import org.gradle.api.internal.tasks.GenericFileNormalizer
+import org.gradle.internal.fingerprint.AbsolutePathInputNormalizer
 import org.gradle.normalization.internal.InputNormalizationStrategy
 import spock.lang.Issue
 
-import static org.gradle.api.internal.changedetection.state.PathNormalizationStrategy.ABSOLUTE
-
 class CacheBackedTaskHistoryRepositoryTest extends AbstractTaskStateChangesTest {
     static final NORMALIZATION_STRATEGY = InputNormalizationStrategy.NOT_CONFIGURED
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy]: I think we should find a better name for this, like `NOT_CONFIGURABLE`.

Comment (wolfs) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy]: How about `NO_NORMALIZATION` or `NOTHING_CONFIGURED`?

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/AbsolutePathInputNormalizer.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/fingerprint/AbsolutePathInputNormalizer.java]: I separated the fingerprinter interfaces from the normalizer interfaces.

I wouldn't make this public API right now, I don't see a reason why someone would want to use those `Normalizer`s. WDYT?

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultAbsolutePathInputNormalizer.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultAbsolutePathInputNormalizer.java]: Done.

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/AbsolutePathInputNormalizer.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/fingerprint/AbsolutePathInputNormalizer.java]: I think it's fine for now.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy (hunk 3 above)
Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy]: `NO_NORMALIZATION` sounds best I think.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ClasspathSnapshotter.java [hunk 4]
```diff
@@ -19,7 +19,7 @@
 import org.gradle.internal.fingerprint.FileCollectionFingerprint;
 
 /**
- * Responsible for calculating a {@link FileCollectionFingerprint} for a {@link org.gradle.api.file.FileCollection} representing a Java classpath. Compared to {@link org.gradle.internal.fingerprint.RelativePathInputNormalizer} this fingerprinter orders files within any sub-tree.
+ * Responsible for calculating a {@link FileCollectionFingerprint} for a {@link org.gradle.api.file.FileCollection} representing a Java classpath. Compared to {@link org.gradle.internal.fingerprint.impl.RelativePathFileCollectionFingerprinter} this fingerprinter orders files within any sub-tree.
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ClasspathSnapshotter.java]: Shouldn't this be in `org.gradle.internal.fingerprint` and called `ClasspathFingerprinter`?

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinter.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.fingerprint.impl;
+
+import org.gradle.api.internal.changedetection.state.FileCollectionSnapshotter;
+
+public interface AbsolutePathFileCollectionFingerprinter extends FileCollectionSnapshotter {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinter.java]: Is it time we renamed `FileCollectionSnapshotter` -> `FileCollectionFingerprinter`, or even `FileFingerprinter`?

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultOutputFileCollectionFingerprinter.java [hunk 6]
```diff
@@ -23,9 +23,9 @@
 import org.gradle.api.tasks.FileNormalizer;
 import org.gradle.internal.fingerprint.OutputNormalizer;
 
-public class DefaultOutputNormalizer extends AbstractPathOnlyFileCollectionFingerprinter implements OutputNormalizer {
+public class DefaultOutputFileCollectionFingerprinter extends AbstractPathOnlyFileCollectionFingerprinter implements OutputFileCollectionFingerprinter {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultOutputFileCollectionFingerprinter.java]: Does it make sense to have a separate `OutputFileCollectionFingerprinter` interface for this? Same for all the other implementations. I think we could simplify the code quite a bit if we discarded the interfaces.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepositoryTest.groovy]: `NO_NORMALIZATION` it is.

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultOutputFileCollectionFingerprinter.java (hunk 6 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultOutputFileCollectionFingerprinter.java]: I removed the interfaces.

**Pull request merged**